src/components/UpvoteButton.js

```diff
@@ -0,0 +1,81 @@
+import * as api from 'api'
+
+import React, { useEffect, useState } from 'react'
+
+import CONSTANTS from 'constants'
+import { Loader } from 'components'
+import PropTypes from 'prop-types'
+import { getCurrentAuthenticatedUser } from 'utility'
+import { isEmpty } from 'lodash'
+import { useNavigate } from 'react-router-dom'
+
+function UpvoteButton({ lightningTalkPollID, numberOfVotes, onClick }) {
+  const [isLoading, setIsLoading] = useState(false)
+  const [buttonText, setButtonText] = useState('Upvote')
+  const navigate = useNavigate()
+
+  useEffect(() => {
+    hasUserVotedBefore()
```

Comment:
This function/API call will run on every button, making many requests. With more time, how would you refactor it to make less requests to the backend?

Reply:
In the current implementation, when the card/button is mounted it makes a fetch request to check whether the current user has upvoted this poll or not. If true, display string value `UNVOTE` to negate the vote, else, display string value `UPVOTE`.

Given more time for better implementation, as you suggest in your question. Assuming we are designing a solution for a huge dataset and a huge user base. I would suggest implementing the following solution:

1. Apply prefetching of the data ahead of time. The place and time to do prefetching depend on the navigation flow. The `home/entry` page and `sign in` page are some common places to do prefetching.

2. Caching the data fetched from the API locally.

3. Syncronyze the locally cached data with the backend. Subscribe to the database's events `onDelete`, `onCreate`, and `onUpdate`.

4. Fetch only the amount of data needed to fill up the user's UI/screen. And fetch the rest gradually.